pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,382 @@
+        """
+        if self.blocking:
+            warnings.warn(
+                "step_and_cost() instead of step() is called when "
+                "blocking is turned on, as the step-wise loss value "
+                "is required by the algorithm.",
+                stacklevel=2,
+            )
+            return self.step_and_cost(cost, *args, **kwargs)[0]
+
+        return self._step_core(cost, args, kwargs)
+
+    def step_and_cost(self, cost, *args, **kwargs):
+        r"""Update trainable parameters with one step of the optimizer and return
+        the corresponding objective function value after the step.
+
+        Args:
+            cost (QNode): the QNode wrapper for the objective function for optimization
+            *args : variable length argument list for qnode
+            **kwargs : variable length of keyword arguments for the qnode
+
+        Returns:
+            (np.array, float): the new variable values :math:`x^{(t+1)}` and the objective
+            function output prior to the step
+        """
+        params_next = self._step_core(cost, args, kwargs)
+
+        if not self.blocking:
+            loss_curr = cost(*args, **kwargs)
+            return params_next, loss_curr
+
+        params_next, loss_curr = self._apply_blocking(cost, args, kwargs, params_next)
+        return params_next, loss_curr
+
+    def _step_core(self, cost, args, kwargs):
+        for i in range(self.resamplings):
+            grad_tapes, grad_dirs = self._get_spsa_grad_tapes(cost, args, kwargs)
+            metric_tapes, tensor_dirs = self._get_tensor_tapes(cost, args, kwargs)
+            raw_results = qml.execute(grad_tapes + metric_tapes, cost.device, None)
+
+            grad = self._post_process_grad(raw_results[:2], grad_dirs)
+            metric_tensor = self._post_process_tensor(raw_results[2:], tensor_dirs)
+
+            if i == 0:
+                grad_avg = grad
+                tensor_avg = metric_tensor
+
+            grad_avg = [grad_avg[j] * i / (i + 1) + grad[j] / (i + 1) for j in range(len(grad))]
+            tensor_avg = tensor_avg * i / (i + 1) + metric_tensor / (i + 1)
+
+        self._update_tensor(tensor_avg)
+        params_next = self._get_next_params(args, grad_avg)
+
+        if len(params_next) == 1:
+            return params_next[0]
+        return params_next
+
+    def _post_process_grad(self, grad_raw_results, grad_dirs):
+        loss_plus, loss_minus = grad_raw_results
+        grad = [
+            (loss_plus - loss_minus) / (2 * self.finite_diff_step) * grad_dir
+            for grad_dir in grad_dirs
+        ]
+        return grad
+
+    def _post_process_tensor(self, tensor_raw_results, tensor_dirs):
+        tensor_raw_results = [result.squeeze() for result in tensor_raw_results]
+
+        tensor_finite_diff = (
+            tensor_raw_results[0][0]
+            - tensor_raw_results[1][0]
+            - tensor_raw_results[2][0]
+            + tensor_raw_results[3][0]
```

Comment:
Maybe comment that these are the probabilities of `\ket{0}`, and hence represent the overlap between the perturbed and unperturbed circuits.